File: RuntimePermissionsRequestJava/app/src/main/java/com/example/runtimepermissionsrequestjava/MainActivity.java
package com.example.runtimepermissionsrequestjava;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import android.Manifest;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.widget.Toast;


public class MainActivity extends AppCompatActivity {
    private static final String PERMISSION_RECORD_AUDIO = Manifest.permission.RECORD_AUDIO;
    private static final int PERMISSION_REQ_CODE = 100;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViewById(R.id.button).setOnClickListener(
                v -> requestRuntimePermission()
        );
    }

    private void requestRuntimePermission() {
        if (ActivityCompat.checkSelfPermission(this, PERMISSION_RECORD_AUDIO)
                == PackageManager.PERMISSION_GRANTED) {
            Toast.makeText(this, "Permission Granted", Toast.LENGTH_SHORT).show();
        } else if (ActivityCompat.shouldShowRequestPermissionRationale(this, PERMISSION_RECORD_AUDIO)) {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setMessage("This app requires RECORD_AUDIO Permission to do sthg")
                    .setTitle("Permission required")
                    .setCancelable(false)
                    .setPositiveButton("OK", (dialogInterface, i) -> {
                ActivityCompat.requestPermissions(MainActivity.this, new String[]{PERMISSION_RECORD_AUDIO}, PERMISSION_REQ_CODE);
                dialogInterface.dismiss();
            }).setNegativeButton("Cancel", ((dialogInterface, i) -> dialogInterface.dismiss()));
            builder.show();
        } else {
            ActivityCompat.requestPermissions(this, new String[]{PERMISSION_RECORD_AUDIO}, PERMISSION_REQ_CODE);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSION_REQ_CODE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(this, "Permission Granted", Toast.LENGTH_SHORT).show();

            } else if (!ActivityCompat.shouldShowRequestPermissionRationale(this, PERMISSION_RECORD_AUDIO)) {
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setMessage("This featue is unvaliverble because you deny ")
                        .setTitle("Permission requried").setCancelable(false)
                        .setNegativeButton("Cancel", ((dialogInterface, i) -> dialogInterface.dismiss()))
                        .setPositiveButton("Settings", (dialogInterface, i) -> {
                            Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                            Uri uri = Uri.fromParts("package", getPackageName(), null);
                            intent.setData(uri);
                            startActivity(intent);

                            dialogInterface.dismiss();
                        });
                builder.show();
            }else{
                requestRuntimePermission();
            }
        }
    }
}

File: RuntimePermissionRequestKotlin/app/src/main/java/com/example/runtimepermissionrequestkotlin/MainActivity.kt
package com.example.runtimepermissionrequestkotlin

import android.Manifest
import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContract
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat

class MainActivity : AppCompatActivity() {

    private lateinit var permisionLauncher: ActivityResultLauncher<Array<String>>
    private var isRecordAudioPermissionGranted = false;
    private var isCameraPermissionGranted = false;
    private var isSenSMSPermissionGranted = false;
    private var isReadSMSPermissionGrangted = false;


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        permisionLauncher = registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()){
            permissions ->
            isRecordAudioPermissionGranted = permissions[Manifest.permission.RECORD_AUDIO]?:isRecordAudioPermissionGranted
            isCameraPermissionGranted = permissions[Manifest.permission.CAMERA]?:isCameraPermissionGranted
            isSenSMSPermissionGranted = permissions[Manifest.permission.SEND_SMS]?:isSenSMSPermissionGranted
            isReadSMSPermissionGrangted = permissions[Manifest.permission.READ_SMS]?:isReadSMSPermissionGrangted
        }
        requestPermission()
        println("ok")
    }
    private fun requestPermission(){
        isRecordAudioPermissionGranted = ContextCompat.checkSelfPermission(
            this,Manifest.permission.RECORD_AUDIO
        ) == PackageManager.PERMISSION_GRANTED

        isCameraPermissionGranted = ContextCompat.checkSelfPermission(
            this, Manifest.permission.CAMERA
        ) == PackageManager.PERMISSION_GRANTED

        isSenSMSPermissionGranted = ContextCompat.checkSelfPermission(
            this, Manifest.permission.SEND_SMS
        ) == PackageManager.PERMISSION_GRANTED

        isReadSMSPermissionGrangted = ContextCompat.checkSelfPermission(
            this, Manifest.permission.READ_SMS
        ) == PackageManager.PERMISSION_GRANTED

        val permissionRequest: MutableList<String> = ArrayList()
        if(!isRecordAudioPermissionGranted){
                permissionRequest.add(Manifest.permission.RECORD_AUDIO)
        }

        if(!isCameraPermissionGranted){
            permissionRequest.add(Manifest.permission.CAMERA)
        }

        if(!isSenSMSPermissionGranted){
            permissionRequest.add(Manifest.permission.SEND_SMS)
        }

        if(!isReadSMSPermissionGrangted){
            permissionRequest.add(Manifest.permission.READ_SMS)
        }


        if(permissionRequest.isNotEmpty()){
            permisionLauncher.launch(permissionRequest.toTypedArray())
        }

        println("ok")
    }

}